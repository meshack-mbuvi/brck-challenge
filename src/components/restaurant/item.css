.details {
    border: 1px solid rgba(0, 0, 0, .125);
    height: 80px;
    margin: 4px 4px 30px 0px;
    display: block;
    float: left;
    background-color: white;
    padding: 2px
}

.card-text {
    padding-bottom: 2px !important;
}

.main {
    background-color: "#f2efef"
}