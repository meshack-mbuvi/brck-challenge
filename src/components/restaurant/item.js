import React from "react";
import StarRatings from "react-star-ratings";
import "./item.css";

/**
 * @description - Ellipsise name if its longer that the 20 characters.
 *
 * @param {string} name - name of the restaurant.
 * @return {string} name - a pre-formated name based on the length of the param.
 */
const formatName = name => {
  if (name.length > 20) {
    return name.slice(0, 20) + "...";
  }

  return name;
};

/**
 * @summary - calculates the average rating for a particular restaurant
 *
 * @param {array} ratings - contains all ratings for the restaurant
 * @return {Number} ratings - average of the ratings provided for a restaurant.
 */
const calculateRating = ratings => {
  const numberOfratings = ratings.length;
  if (!numberOfratings) return 0;

  const sum = ratings.reduce((acc, currentValue) => acc + currentValue);
  return sum / numberOfratings;
};

/**
 * @summary - component to display a single restaurant.
 *
 * @param {Object} props
 * @return {DOMnode}
 */
export const Restaurant = props => {
  const { name, ratings } = props.data;
  const rating = calculateRating(ratings);

  return (
    <div className="details card col-md-3">
      <a href={name} className="card-link">
        {formatName(name)}
      </a>

      <div className="card-footer">
        <span>Ratings: </span>
        <StarRatings
          rating={rating}
          starRatedColor="blue"
          numberOfStars={5}
          name="rating"
          starDimension="10px"
          starSpacing="2px"
        />
      </div>
    </div>
  );
};

export default Restaurant;
