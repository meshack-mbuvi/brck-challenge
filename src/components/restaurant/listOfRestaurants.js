import React, { Component } from "react";
import { Restaurant } from "./item";
import "./item.css";

export class ListOfRestaurants extends Component {
  render() {
    const { restaurants } = this.props;

    const items = restaurants
      ? restaurants
          .slice(0, 12)
          .map((restaurant, index) => (
            <Restaurant data={restaurant} key={index} />
          ))
      : restaurants;

    return <div className="col-md-12">{items}</div>;
  }
}

export default ListOfRestaurants;
