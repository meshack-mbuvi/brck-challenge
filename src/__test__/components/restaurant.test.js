import React from "react";

import Enzyme, { shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import expect from "expect";

import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import { ListOfRestaurants } from "../../components/restaurant/listOfRestaurants";
import Restaurant from "../../components/restaurant/item";
import Pagination from "../../components/pagination";
import Filter from "../../components/filter";

Enzyme.configure({ adapter: new Adapter() });

const middleware = [thunk];
const mockStore = configureStore(middleware);
const store = mockStore({});

const props = {
  store,
  dispatch: jest.fn(),
  restaurants: [
    {
      name: "Kushi Tsuru",
      ratings: [],
      opening_time: ["Mon-Sun 11 am - 10 pm"],
    },
  ],
  handlePageChange: jest.fn(),
  currentPage: 1,
  limitPerPage: 10,
  handleOptionChange: jest.fn(),
};

describe("Restaunts", () => {
  let wrapper = shallow(<ListOfRestaurants props={props} />);

  it("renders <ListOfRestaurant/> correctly", () => {
    expect(wrapper.find(".col-md-12").length).toEqual(1);
  });

  it("renders <Restaurant/> component", () => {
    wrapper = shallow(<Restaurant data={props.restaurants[0]} />);
    expect(wrapper.find(".details").length).toEqual(1);
  });

  it("renders <Pagination/> correctly", () => {
    wrapper = shallow(<Pagination {...props} />);
    expect(wrapper).toMatchSnapshot();

    wrapper.find("#paginationPrevious").simulate("click");
    expect(props.handlePageChange).toBeCalledTimes(1);
  });

  it("renders <Filter/> correctly", () => {
    wrapper = shallow(<Filter {...props} />);
    expect(wrapper).toMatchSnapshot();

    wrapper.find("select").simulate("change", { target: { value: "open" } });
    expect(props.handleOptionChange).toBeCalledTimes(1);
  });
});
