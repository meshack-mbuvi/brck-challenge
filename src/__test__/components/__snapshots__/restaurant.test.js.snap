// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Restaunts renders <Filter/> correctly 1`] = `ShallowWrapper {}`;

exports[`Restaunts renders <Pagination/> correctly 1`] = `ShallowWrapper {}`;
